yozhao101/sonic-buildimage
# [Service] Restart BGP service upon unexpected critical process exited.

- What I did
Restart BGP service if one of critical processes running in teamd container exited abnormally.

- How I did it
Generally I follow the framework created by Joe to implement this feature in BGP container.
**First**, add supervisor-proc-exit-listener event listener plugin for Supervisor in Teamd Docker
container which in turn loads a list of critical processes for which to monitor for unexpected
exits.
**Second**, configure temad.service to always auto-restart the service if it stops, with a delay of 30 seconds. Also set a rate limit of 3 restarts within 20 minutes (1200 seconds).

- How to verify it
On your switch device, please use docker ps command to list the running docker containers.
Then use docker exec -it container_id /bin/bash to login your container. Type top command
on the shell will display all the processes dynamically and you will spot the process id of one
of the critical processes. Finally type the command kill -9 process_id to kill one process.
After exiting the container, you can use watch -n 1 docker ps to dynamically see the restart
of teamd and dhcp-relay containers.
